src/libtoast/src/toast_tod_filter.cpp

```diff
@@ -52,108 +58,228 @@ void toast::filter_polynomial(int64_t order, size_t n, uint8_t * flags,
+                }
+            }
 
-            for (size_t isignal = 0; isignal < nsignal; ++isignal) {
-                double * signal = signals[isignal] + start;
+            // Fit the templates against the data.  DGELSS uses SVD to
+            // minimize the norm of the difference and the solution
+            // vector.
+
+            toast::AlignedVector <double> singular_values(norder);
+            int rank, info;
+            double rcond_limit = 1e-3;
+            int lwork = std::max(10 * (norder + nsignal), 1000000);
+            toast::AlignedVector <double> work(lwork);
+
+            // DGELSS will overwrite proj with the fitting
+            // coefficients.  invcov is overwritten with
+            // singular vectors.
+            toast::lapack_dgelss(&norder, &norder, &nsignal,
+                                 invcov.data(), &norder,
+                                 proj.data(), &norder,
+                                 singular_values.data(), &rcond_limit,
+                                 &rank, work.data(), &lwork, &info);
+
+            for (int iorder = 0; iorder < norder; ++iorder) {
+                double * temp = &full_templates[iorder * scanlen];
+                for (int isignal = 0; isignal < nsignal; ++isignal) {
+                    double * signal = &signals[isignal][start];
+                    double amp = proj[iorder + isignal * norder];
+                    #pragma omp simd
+                    for (size_t i = 0; i < scanlen; ++i) signal[i] -= amp * temp[i];
+                }
+            }
+        }
+    }
 
-                // proj = templates x signal
+    return;
+}
 
-                toast::lapack_gemv(&notrans, &fnorder, &fscanlen, &fone,
-                                   templates.data(), &fnorder, signal, &one,
-                                   &fzero, proj.data(), &one);
+void toast::bin_templates(double * signal, double * templates,
+                          uint8_t * good, double * invcov, double * proj,
+                          size_t nsample, size_t ntemplate) {
+    for (size_t row = 0; row < ntemplate; row++) {
+        proj[row] = 0;
+        for (size_t col = 0; col < ntemplate; col++) {
+            invcov[ntemplate * row + col] = 0;
+        }
+    }
 
-                // coeff = cov x proj
+#pragma omp parallel for \
+    schedule(static) default(none) shared(proj, templates, signal, good, ntemplate, nsample)
+    for (size_t row = 0; row < ntemplate; ++row) {
+        double * ptemplate = templates + row * nsample;
+        for (size_t i = 0; i < nsample; ++i) {
+            proj[row] += ptemplate[i] * signal[i] * good[i];
+        }
+    }
 
-                // For whatever reason, trmv refused to yield the right
-                // answer...
-                // Use general matrix vector multiply instead
+#pragma omp parallel \
+    default(none) shared(invcov, templates, signal, good, ntemplate, nsample)
+    {
+        int nthread = omp_get_num_threads();
```

Comment:
These calls to `omp_*` functions need to be guarded like:
```
int nthread = 1;
int id_thread = 0;
#ifdef _OPENMP
nthread = omp_get_num_threads();
id_thread = omp_get_thread_num();
#endif
```

Reply:
Done.